Implement aspect-ratio mapping for canvas element.

sammygill opened the pull request:
#### 8c0165e26f14260d13eabf2d7c8ca675f3bc0197
<pre>
Implement aspect-ratio mapping for canvas element.
<a href="https://bugs.webkit.org/show_bug.cgi?id=240887">https://bugs.webkit.org/show_bug.cgi?id=240887</a>

Reviewed by Chris Dumez.

This patch adds the mapping between the width and height attributes,
which are both presentational hints, to the aspect-ratio property on
canvas elements. When a canvas element is created, its own
implementations of hasPresentationalHintsForAttribute and
collectPresentationalHintsForAttribute will be used to set the intrinsic
width and height of the element as well as set the aspect-ratio property
if both are provided.

<a href="https://html.spec.whatwg.org/multipage/rendering.html#attributes-for-embedded-content-and-images">https://html.spec.whatwg.org/multipage/rendering.html#attributes-for-embedded-content-and-images</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/replaced-elements/attributes-for-embedded-content-and-images/canvas-aspect-ratio-expected.txt:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::hasPresentationalHintsForAttribute const):
(WebCore::HTMLCanvasElement::collectPresentationalHintsForAttribute):
* Source/WebCore/html/HTMLCanvasElement.h:

* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::addParsedWidthAndHeightToAspectRatioList):
(WebCore::HTMLElement::applyAspectRatioFromWidthAndHeightAttributesToStyle):
(WebCore::HTMLElement::applyAspectRatioWithoutDimensionalRulesFromWidthAndHeightAttributesToStyle):
There are two different ways to map the width and height to the aspect ratio:
1.) map to the aspect-ratio property (using dimension rules)
2.) map to the aspect-ratio property

The main difference in the parsing process is that only integer values
are outputted when parsing without dimension rules and dimension rules
will output any sort of number. For example, 0.5 will be parsed as 0.5
when usig 1, but will be 0 when parsed with 2. This new method parses
the width and height without dimensional rules and uses the results
for the aspect-ratio property.

<a href="https://html.spec.whatwg.org/multipage/rendering.html#map-to-the-aspect-ratio-property">https://html.spec.whatwg.org/multipage/rendering.html#map-to-the-aspect-ratio-property</a>

* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/rendering/RenderBox.h:
(WebCore::RenderBox::isAspectRatioDegenerate const):
The CSS spec defines a degenerate aspect-ratio as one that is either
0 or infinite. This helper method can be used in places where we
perform an if(aspectRatio) check like in RenderReplaced::computeReplacedLogicalWidth
so that we can catch both of the cases. Before adding this method, the
case where the aspect ratio is infinite would fall through into the
if branch when it should not have.

* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::computeReplacedLogicalWidth const):

Canonical link: <a href="https://commits.webkit.org/252001@main">https://commits.webkit.org/252001@main</a>
</pre>

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -521,6 +521,28 @@ void HTMLElement::applyAspectRatioFromWidthAndHeightAttributesToStyle(StringView
+
+    style.setProperty(CSSPropertyAspectRatio, RefPtr<CSSValue>(WTFMove(list)));
+}
+
```
I looked into how to not duplicate most of this function. Seems like places that use the "dimension rule" we also parse a dimension and map the width/height. For `<canvas>` (the only place not using that rule), we don't map those attributes. 

I wonder if we can leave the parsing bits to wherever width/height are parsed for places using the dimension rule, and make this function its own (taking in a width & height):

```
    auto ratioList = CSSValueList::createSlashSeparated();
      ratioList->append(CSSValuePool::singleton().createValue(dimensionWidth.value(), CSSUnitType::CSS_NUMBER));
      ratioList->append(CSSValuePool::singleton().createValue(dimensionHeight.value(), CSSUnitType::CSS_NUMBER));
      auto list = CSSValueList::createSpaceSeparated();
      list->append(CSSValuePool::singleton().createIdentifierValue(CSSValueAuto));
      list->append(ratioList);
 
      style.setProperty(CSSPropertyAspectRatio, RefPtr<CSSValue>(WTFMove(list)));
```

For `<canvas>`, you'd call `parseHTMLNonNegativeInteger` then pass the width/height onto that new function.
 
You can also check out Firefox's implementation: https://phabricator.services.mozilla.com/D109618

sammygill replied:
I actually think that not mapping the width/height in this case is an error on my end and is causing some tests to fail. I'll be taking a look into that to see if it resolves the issue, but I still agree with what you are saying about the rest of it and will see what I can do!

nt1m replied:
I don't think canvas needs to have the width/height attributes mapped to width/height CSS props, if I'm reading the spec correctly: https://html.spec.whatwg.org/multipage/rendering.html#attributes-for-embedded-content-and-images:map-to-the-aspect-ratio-property

nt1m replied:
It's only for this subset of elements: https://html.spec.whatwg.org/multipage/rendering.html#attributes-for-embedded-content-and-images:maps-to-the-dimension-property

sammygill commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -511,9 +511,29 @@ void HTMLElement::applyAspectRatioFromWidthAndHeightAttributesToStyle(StringView
+    if (!dimensionHeight)
+        return;
+
+    addParsedWidthAndHeightToAspectRatioList(dimensionWidth.value(), dimensionHeight.value(), style);
```
I added this method to help deal with the duplication issue that we were talking about. What do you think of this as a solution? @nt1m

cdumez commented on Source/WebCore/html/HTMLCanvasElement.h:
```diff
@@ -153,6 +153,8 @@ class HTMLCanvasElement final : public HTMLElement, public CanvasBase, public Ac
     void eventListenersDidChange() final;
 
     void parseAttribute(const QualifiedName&, const AtomString&) final;
+    bool hasPresentationalHintsForAttribute(const QualifiedName&) const override;
```
nit: Would be nice to mark as final

sammygill replied:
Made the change. Methods should now be final

cdumez commented on Source/WebCore/html/HTMLCanvasElement.h:
```diff
@@ -153,6 +153,8 @@ class HTMLCanvasElement final : public HTMLElement, public CanvasBase, public Ac
 
     void parseAttribute(const QualifiedName&, const AtomString&) final;
+    bool hasPresentationalHintsForAttribute(const QualifiedName&) const override;
+    void collectPresentationalHintsForAttribute(const QualifiedName&, const AtomString&, MutableStyleProperties&) override;
```
ditto

sammygill replied:
Made the change. Methods should now be final

cdumez commented on Source/WebCore/rendering/RenderBox.h:
```diff
@@ -731,7 +731,8 @@ override;
-
+    
+    bool isAspectRatioDegenerate(double aspectRatio) const { return (!aspectRatio || isnan(aspectRatio)); }
 private:
```
nit: Could you please keep the blank line before `private:`?

cdumez replied:
Also, the parentheses are unnecessary:
`bool isAspectRatioDegenerate(double aspectRatio) const { return !aspectRatio || isnan(aspectRatio); }`

sammygill replied:
Fixed!

cdumez reviewed (approved):
LGTM

sammygill committed:
Implement aspect-ratio mapping for canvas element.
https://bugs.webkit.org/show_bug.cgi?id=240887

Reviewed by Chris Dumez.

This patch adds the mapping between the width and height attributes,
which are both presentational hints, to the aspect-ratio property on
canvas elements. When a canvas element is created, its own
implementations of hasPresentationalHintsForAttribute and
collectPresentationalHintsForAttribute will be used to set the intrinsic
width and height of the element as well as set the aspect-ratio property
if both are provided.

https://html.spec.whatwg.org/multipage/rendering.html#attributes-for-embedded-content-and-images

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/replaced-elements/attributes-for-embedded-content-and-images/canvas-aspect-ratio-expected.txt:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::hasPresentationalHintsForAttribute const):
(WebCore::HTMLCanvasElement::collectPresentationalHintsForAttribute):
* Source/WebCore/html/HTMLCanvasElement.h:

* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::addParsedWidthAndHeightToAspectRatioList):
(WebCore::HTMLElement::applyAspectRatioFromWidthAndHeightAttributesToStyle):
(WebCore::HTMLElement::applyAspectRatioWithoutDimensionalRulesFromWidthAndHeightAttributesToStyle):
There are two different ways to map the width and height to the aspect ratio:
1.) map to the aspect-ratio property (using dimension rules)
2.) map to the aspect-ratio property

The main difference in the parsing process is that only integer values
are outputted when parsing without dimension rules and dimension rules
will output any sort of number. For example, 0.5 will be parsed as 0.5
when usig 1, but will be 0 when parsed with 2. This new method parses
the width and height without dimensional rules and uses the results
for the aspect-ratio property.

https://html.spec.whatwg.org/multipage/rendering.html#map-to-the-aspect-ratio-property

* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/rendering/RenderBox.h:
(WebCore::RenderBox::isAspectRatioDegenerate const):
The CSS spec defines a degenerate aspect-ratio as one that is either
0 or infinite. This helper method can be used in places where we
perform an if(aspectRatio) check like in RenderReplaced::computeReplacedLogicalWidth
so that we can catch both of the cases. Before adding this method, the
case where the aspect ratio is infinite would fall through into the
if branch when it should not have.

* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::computeReplacedLogicalWidth const):

Canonical link: https://commits.webkit.org/252001@main
